### PR TITLE
Add encodeEncodedWords function for creating headers with non-ASCII

### PR DESCRIPTION
**Pull request by singpolyma**
Takes in the Text that you want in the header and returns either the
same text as an ASCII ByteString or an appropriate encoded-word that
will decode to the original.

### CODE REVIEW

File: src/Data/MIME/EncodedWord.hs [hunk 1]
```diff
@@ -144,3 +146,17 @@ decodeEncodedWords charsets s =
...
+-- Currently turns the whole text into a single encoded word, if necessary.
+encodeEncodedWords :: T.Text -> B.ByteString
+encodeEncodedWords s
+  | cteName == "7bit" = s'
+  | cteName == "base64" = "=?utf-8?B?" <> s' <> "?="
+  | cteName == "quoted-printable" = "=?utf-8?Q?" <> s' <> "?="
+  | otherwise = "=?utf-8?X?" <> s' <> "?="
+  where
+    s' = C.filter (`notElem` ['\r', '\n']) $ review (clonePrism cte) bytes
+    (cteName, cte) = chooseTransferEncoding bytes
```
Comment (frasertweedale) [inline src/Data/MIME/EncodedWord.hs]: Thanks for the PR.  The 'Q' encoding is not quite the same as Quoted-Printable.  So we ought to have a dedicated function `chooseEncodedWordEncoding :: T.Text -> Maybe (TransferEncodingName, TransferEncoding)` that provides the `B` or `Q` encoding prism if encoding is needed.  At the call site, for the `Just` case, construct an `EncodedWord` and apply `serialiseEncodedWord` to it.

The tests look good.

Comment (frasertweedale) [reply src/Data/MIME/EncodedWord.hs]: I'm happy to leave these changes to you, or I can take it from here if you wish.

Comment (singpolyma) [reply src/Data/MIME/EncodedWord.hs]: I've made these changes now, I think.

Comment (frasertweedale) [reply src/Data/MIME/EncodedWord.hs]: @singpolyma thanks, I'll review it in the next day or so.

**Comment by frasertweedale** [comment]
@singpolyma I have a general question.  What is your use case?  i.e. in what context are you turning Text into ByteString?

As part of fix for #50 I now have a function `buildPhrase :: T.Text -> B.Builder`.  Note some productions from RFC 5322 and RFC 2047:

```
   phrase = 1*( encoded-word / word )
   word            =   atom / quoted-string
   atom            =   [CFWS] 1*atext [CFWS]
```

`buildPhrase` uses *atom* if possible, otherwise *quoted-string* if possible, otherwise *encoded-word*.  Is it suitable for your use case?

**Comment by singpolyma** [comment]
My *main* usecase is encoding the `Subject` header, so I wouldn't want it to be wrapped in quotes.

**Comment by frasertweedale** [comment]
@singpolyma thanks for clarifying.  I might introduce a specific optic to expose the subject header as 'Maybe Text', since it seems like a pretty common need :)

**Comment by frasertweedale** [comment]
@singpolyma I'm sorry it has been a while.  I have made a few enhancements, including encoding only those words that require it rather than the whole input.  Are you happy with what's in https://github.com/purebred-mua/purebred-email/tree/fix/50-mailbox-display-name-unicode ?

**Comment by singpolyma** [comment]
The comment:

     -- Currently turns the whole text into a single encoded word, if necessary.

is wrong now, but otherwise seems good (passes the tests in my application anyway, with appropriate changes).

**Comment by frasertweedale** [comment]
@singpolyma thanks for testing, and for picking up on the haddock needing fixing :+1:

**Comment by frasertweedale** [comment]
This was merged separately from the PR.

**Pull request closed without merging**